libnopegl/src/text.c

```diff
@@ -127,6 +138,220 @@ int ngli_text_init(struct text *s, const struct text_config *cfg)
+
+        if (c->tags & mask) {
+            if (inside_element) {
+                position++;
+                inside_element = 0;
+            }
+        } else if (!inside_element) {
+            effect->total_segments++;
+            inside_element = 1;
+        }
+
+        if (!(c->tags & NGLI_TEXT_CHAR_TAG_GLYPH))
+            continue;
+
+        effect->positions[char_id++] = position;
+    }
+    ngli_assert(char_id == ngli_darray_count(&s->chars));
+}
+
+static void segment_words(const struct text *s, struct effect_segmentation *effect)
+{
+    segment_separator(s, effect, NGLI_TEXT_CHAR_TAG_WORD_SEPARATOR);
+}
+
+static void segment_lines(const struct text *s, struct effect_segmentation *effect)
+{
+    segment_separator(s, effect, NGLI_TEXT_CHAR_TAG_LINE_BREAK);
+}
+
+static void segment_text(const struct text *s, struct effect_segmentation *effect)
+{
+    for (size_t i = 0; i < ngli_darray_count(&s->chars); i++)
+        effect->positions[i] = 0;
+    effect->total_segments = 1;
+}
+
+/*
+ * SplitMix64, public domain code from Sebastiano Vigna (2015)
+ * See https://xoshiro.di.unimi.it/splitmix64.c
+ */
+static uint64_t prng64_next(uint64_t *state)
+{
+    uint64_t z = *state;
+    *state += 0x9e3779b97f4a7c15;
+    z = (z ^ (z >> 30)) * 0xbf58476d1ce4e5b9;
+    z = (z ^ (z >> 27)) * 0x94d049bb133111eb;
+    return z ^ (z >> 31);
+}
+
+static void shuffle(uint64_t *rng_state, size_t *positions, size_t n)
+{
+    for (size_t i = 0; i < n - 1; i++) {
+        const size_t r = i + (size_t)prng64_next(rng_state) % (n - i);
+        NGLI_SWAP(size_t, positions[i], positions[r]);
+    }
+}
+
+static int build_effects_segmentation(struct text *s)
+{
+    for (size_t i = 0; i < s->config.nb_effect_nodes; i++) {
+        struct ngl_node *effect_node = s->config.effect_nodes[i];
+        struct texteffect_opts *effect_opts = effect_node->opts;
+        struct effect_segmentation *effect = &s->effects[i];
+
+        ngli_freep(&effect->positions);
+
+        const size_t nb_chars = ngli_darray_count(&s->chars);
+        effect->positions = ngli_calloc(nb_chars, sizeof(*effect->positions));
+        if (!effect->positions)
+            return NGL_ERROR_MEMORY;
+
+        switch (effect_opts->target) {
+        case NGLI_TEXT_EFFECT_CHAR:         segment_chars(s, effect);         break;
+        case NGLI_TEXT_EFFECT_CHAR_NOSPACE: segment_chars_nospace(s, effect); break;
+        case NGLI_TEXT_EFFECT_WORD:         segment_words(s, effect);         break;
+        case NGLI_TEXT_EFFECT_LINE:         segment_lines(s, effect);         break;
+        case NGLI_TEXT_EFFECT_TEXT:         segment_text(s, effect);          break;
+        default:
+            ngli_assert(0);
+        }
+
+        if (effect_opts->random) {
+            /* Build a shuffle map associating a position with another one */
+            size_t *shuffle_map = ngli_calloc(effect->total_segments, sizeof(*shuffle_map));
```

Comment:
Same here ?

Reply:
Mmh this one is local, not sure it's worth keeping it in the context

Reply:
The idea is to lower the number of alloc on text updates. It could be left as-is though, we still have the option to rework it later on.